src/action/file_stat.rs

```diff
@@ -0,0 +1,532 @@
+}
+
+#[derive(Debug)]
+struct PathSpec {
+    nested_path: Option<Box<PathSpec>>,
+    path: Option<PathBuf>,
+}
+
+impl Default for PathSpec {
+    fn default() -> PathSpec {
+        PathSpec {
+            nested_path: None,
+            path: None,
+        }
+    }
+}
+
+impl Default for Response {
+    fn default() -> Response {
+        Response {
+            mode: None,
+            inode: None,
+            device: None,
+            hard_links: None,
+            uid: None,
+            gid: None,
+            size: None,
+            access_time: None,
+            modification_time: None,
+            status_change_time: None,
+            blocks_number: None,
+            block_size: None,
+            represented_device: None,
+            flags_linux: None,
+            symlink: None,
+            pathspec: PathSpec::default(),
+            extended_attributes: vec![],
+        }
+    }
+}
+
+pub fn handle<S: Session>(session: &mut S, request: Request) -> session::Result<()> {
+    let original_path = collapse_pathspec(request.pathspec.unwrap());
+
+    let follow_symlink = match request.follow_symlink {
+        Some(s) => s,
+        None => false,
+    };
+
+    let collect_ext_attrs = match request.collect_ext_attrs {
+        Some(s) => s,
+        None => false,
+    };
+
+    let destination = if follow_symlink {
+        fs::canonicalize(&original_path)?
+    } else {
+        original_path.clone()
+    };
+
+    let mut response = form_response(&original_path, &destination)?;
+    if collect_ext_attrs {
+        response.extended_attributes = get_ext_attrs(&destination);
+    }
+
+    session.reply(response)?;
+    Ok(())
+}
+
+#[cfg(target_family = "unix")]
+fn get_ext_attrs(path: &Path) -> Vec<ExtAttr> {
+    let xattrs = xattr::list(path).unwrap();
+
+    let mut result = vec![];
+    for attr in xattrs {
+        result.push(ExtAttr {
+            name: Some(attr.to_str().unwrap().as_bytes().to_vec()),
+            value: xattr::get(path, attr).unwrap(),
+        });
+    }
+    result
+}
+
+#[cfg(not(target_family = "unix"))]
+fn get_ext_attrs(path: &Path) -> Vec<ExtAttr> {
+    vec![]
+}
+
+fn get_time_option<E: std::fmt::Display>(time: Result<SystemTime, E>) -> Option<SystemTime> {
+    match time {
+        Ok(time_value) => Some(time_value),
+        Err(err) => {
+            warn!("Unable to get time value: {}", err);
+            None
+        }
+    }
+}
+
+fn get_time_since_unix_epoch(sys_time: &Option<SystemTime>) -> Option<u64> {
+    return sys_time.map_or(None, |time| time.duration_since(UNIX_EPOCH)
+        .map_or(None, |dur| Some(dur.as_secs())));
+}
+
+#[cfg(target_os = "linux")]
+fn get_status_change_time(metadata: &Metadata) -> Option<SystemTime> {
+    use std::time::Duration;
+    use std::os::unix::fs::MetadataExt;
+
+    UNIX_EPOCH.checked_add(Duration::from_secs(metadata.ctime() as u64))
+}
+
+#[cfg(target_os = "linux")]
+fn form_response(original_path: &PathBuf, destination: &PathBuf)
```

Comment:
Passing `PathBuf` as a reference is an anti-pattern, use `Path` or `AsRef<Path>` instead (take hint from [the standard library](https://doc.rust-lang.org/std/fs/fn.metadata.html)).

Reply:
Fixed.